osbuild/monitor.py

```diff
@@ -0,0 +1,134 @@
+
+
+RESET = "\033[0m"
+BOLD = "\033[1m"
+
+RECORD_SEPARATOR = u"\u241E"
+LINE_FEED = u"\u000A"
+
+
+class BaseMonitor:
+    def begin(self, pipeline: osbuild.Pipeline):
+        """Called once at the beginning of a build"""
+
+    def finish(self, result: Dict):
+        """Called at the very end of the build"""
+
+    def stage(self, stage: osbuild.Stage):
+        """Called when a stage is being built"""
+
+    def assembler(self, assembler: osbuild.Assembler):
+        """Called when an assembler is being built"""
+
+    def result(self, result: osbuild.pipeline.BuildResult):
+        """Called when a module is done with its result"""
+
+    def log(self, message: str):
+        """Called for all module log outputs"""
+
+
+class FileMonitor(BaseMonitor):
+    def __init__(self, file):
+        self.out = file
+
+    def stage(self, stage):
+        self.module(stage)
+
+    def assembler(self, assembler):
+        self.out.write("{RESET}{BOLD}Assembler{RESET} ")
+        self.module(assembler)
+
+    def module(self, module):
+        title = f"{module.name}: {module.id}"
+        options = module.options or {}
+
+        self.out.write(f"{RESET}{BOLD}{title}{RESET} ")
+        json.dump(options, self.out, indent=2)
+        self.out.write("\n")
+
+    def log(self, message):
+        self.out.write(message)
+
+
+class JsonStream(BaseMonitor):
+
+    def __init__(self, file):
+        self.out = file
+        self.modules = {}
+
+    def begin(self, pipeline):
+
+        def register(module):
+            count = len(self.modules) + 1
+            self.modules[module.id] = count
+
+        def collect(pl):
+            if pl.build:
+                collect(pl.build)
+
+            for stage in pl.stages:
+                register(stage)
+
+            if pl.assembler:
+                register(pl.assembler)
+
+        collect(pipeline)
+
+    def stage(self, stage):
+        self.module(stage, "stage")
+
+    def assembler(self, assembler):
+        self.module(assembler, "assembler")
+
+    def module(self, module, module_type):
+        module_id = module.id
+        number = self.modules[module_id]
+
+        status = {
+            "type": "build",
+            "module": {
+                "type": module_type,
+                "name": module.name,
+                "id": module_id
+            },
+            "progress": {
+                "current": number,
+                "total": len(self.modules)
+            }
+        }
+
+        self.send(status)
+
+    def result(self, result):
+        status = {
+            "type": "result",
+            "module": {
+                "name": result.name,
+                "id": result.id,
+            },
+            "success": result.success,
+            "output": result.output
+        }
+
+        self.send(status)
+
+    def send(self, js):
+        sys.stderr.write(RECORD_SEPARATOR)
+        json.dump(js, self.out, indent=2)
+        sys.stderr.write(LINE_FEED)
+
+
+def for_args(args):
```

Comment:
Why pass all of args instead of just a bool?

Reply:
Ah, indeed, I had `interactive: bool` before, but I think the future there will be `verbose` and `streaming` and such and I didn't want to create a flags monster. But maybe I should cross that bridge once there are too many of those.

Reply:
Yeah, the tradeoff is that it's harder to document and check the args (eg. linting, type checking) when everything is buried in the object. But on the other hand when everything is an argument you end up with giant sprawling argument lists like some of the functions in lorax :/ I haven't figured out a good middle ground for lots of args, but for a few I think it's important to just pass them.

Reply:
Fair point, I think I am convinced. Maybe if it gets out of hand we need a `MonitorOptions` `namedtuple` or something.

Reply:
Done! Thanks!